File: frontend/src/utils/autolink.ts
const EMAIL_RE = /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/g
const USERNAME_RE = /(?<![\w@])(@[A-Za-z0-9_]{1,32})(?![\w@])/g
const URL_RE = /(?:https?:\/\/|www\.)[^\s<>"'`]+/gi
const ESCAPE_RE = /[&<>"']/g

function escapeHtml(text: string): string {
  const map: Record<string, string> = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#39;',
  }
  return text.replace(ESCAPE_RE, (ch) => map[ch])
}

function displayDomain(url: string): string {
  try {
    const u = new URL(url.startsWith('http') ? url : `https://${url}`)
    return u.hostname.replace(/^www\./, '')
  } catch {
    return url.replace(/^https?:\/\//, '').replace(/^www\./, '')
  }
}

/** Автоссылки на URL, email и @username в тексте. Возвращает безопасный HTML. */
export function autolinkText(text: string): string {
  const placeholders: string[] = []

  const withUrls = text.replace(URL_RE, (url) => {
    const href = url.startsWith('http') ? url : `https://${url}`
    const label = displayDomain(url)
    placeholders.push(
      `<a href="${escapeHtml(href)}" class="msg-link" target="_blank" rel="noopener noreferrer">${escapeHtml(label)}</a>`,
    )
    return `\u0000${placeholders.length - 1}\u0000`
  })

  const escaped = escapeHtml(withUrls)
  const emailLinked = escaped.replace(EMAIL_RE, (email) => {
    const href = `mailto:${email}`
    return `<a href="${href}" class="underline" target="_blank" rel="noopener noreferrer">${email}</a>`
  })
  const usernameLinked = emailLinked.replace(USERNAME_RE, (username) => {
    const slug = username.slice(1)
    const href = `https://t.me/${slug}`
    return `<a href="${href}" class="underline" target="_blank" rel="noopener noreferrer">${username}</a>`
  })

  // \u0000 используется как безопасный sentinel-разделитель для URL-плейсхолдеров
  // eslint-disable-next-line no-control-regex
  return usernameLinked.replace(/\u0000(\d+)\u0000/g, (_m, i: string) => placeholders[Number(i)] ?? '')
}

File: backend/app/utils/autolink.py
import html
import re
from typing import Optional
from urllib.parse import urlparse

_URL_RE = re.compile(r"(?:https?://|www\.)[^\s<>\"']+", re.IGNORECASE)


def _display_domain(url: str) -> str:
    try:
        parsed = urlparse(url if "://" in url else "https://" + url)
        host = parsed.hostname or url
        return host.removeprefix("www.")
    except Exception:
        return url


def autolink(content: Optional[str]) -> str:
    """Экранирует текст и превращает http/https/www URL в кликабельные ссылки.

    Показывается домен (а не полный URL), ссылка помечается классом msg-link
    (синяя → фиолетовая после клика, стили во фронте).
    """
    if not content:
        return ""
    escaped = html.escape(content, quote=True)
    return _URL_RE.sub(_wrap_url, escaped)


def _wrap_url(match: "re.Match[str]") -> str:
    url = match.group(0)
    href = url if url.lower().startswith(("http://", "https://")) else "https://" + url
    label = html.escape(_display_domain(url))
    return (
        f'<a href="{href}" class="msg-link" target="_blank" rel="noopener noreferrer">'
        f"{label}</a>"
    )
